src/battle_util.c

```diff
@@ -9622,6 +9622,21 @@ void UndoFormChange(u32 monId, u32 side, bool32 isSwitchingOut)
             break;
         }
     }
+    if (!isSwitchingOut)
+    {
+        // Apply party-wide end-of-battle form changes
+        for (i = 0; i < PARTY_SIZE; i++)
+        {
+            // Player's side
+            targetSpecies = GetFormChangeTargetSpecies(&gPlayerParty[i], FORM_BATTLE_END, 0);
+            if (targetSpecies != SPECIES_NONE)
+                SetMonData(&gPlayerParty[i], MON_DATA_SPECIES, &targetSpecies);
+            // Opponent's side
+            targetSpecies = GetFormChangeTargetSpecies(&gEnemyParty[i], FORM_BATTLE_END, 0);
+            if (targetSpecies != SPECIES_NONE)
+                SetMonData(&gEnemyParty[i], MON_DATA_SPECIES, &targetSpecies);
+        }
+    }
```

Comment:
Putting this inside of `UndoFormChange` will force this change for *all mon in both parties* in all the places where it's not switching out, which is most of them. Eg: reverting Mimikyu in when a pokemon faints.
I think the correct approach is to get rid of `UndoFormChange` entirely, have a new macro that handles form change and use it where necessary.

Reply:
That's exactly what I wanted to do yesterday when I talked to you about a `FORM_BATTLE_HIT` 😅